packages/shared/src/enums.ts

```diff
@@ -26,71 +30,130 @@ export type ChainflipNetwork =
+    decimals: 18,
+    contractId: 2,
+  },
+  [InternalAssets.Usdc]: {
+    chain: Chains.Ethereum,
+    asset: Assets.USDC,
+    name: 'USDC',
+    decimals: 6,
+    contractId: 3,
+  },
+  [InternalAssets.Dot]: {
+    chain: Chains.Polkadot,
+    asset: Assets.DOT,
+    name: 'Polkadot',
+    decimals: 10,
+    contractId: 4,
+  },
+  [InternalAssets.Btc]: {
+    chain: Chains.Bitcoin,
+    asset: Assets.BTC,
+    name: 'Bitcoin',
+    decimals: 8,
+    contractId: 5,
+  },
+} satisfies Record<
+  InternalAsset,
+  {
+    chain: Chain;
+    asset: Asset;
+    name: string;
+    decimals: number;
+    contractId: number;
+  }
+>;
 
 // https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
-export const chainContractIds: Record<Chain, number> = {
-  [Chains.Ethereum]: 1,
-  [Chains.Polkadot]: 2,
-  [Chains.Bitcoin]: 3,
-};
-
+export const chainConstants = {
+  [Chains.Ethereum]: {
+    assets: [Assets.ETH, Assets.FLIP, Assets.USDC],
+    gasAsset: Assets.ETH,
+    contractId: 1,
+  },
+  [Chains.Polkadot]: {
+    assets: [Assets.DOT],
+    gasAsset: Assets.DOT,
+    contractId: 2,
+  },
+  [Chains.Bitcoin]: {
+    assets: [Assets.BTC],
+    gasAsset: Assets.BTC,
+    contractId: 3,
+  },
+} satisfies Record<
+  Chain,
+  { assets: Asset[]; gasAsset: Asset; contractId: number }
+>;
+
+export type AssetOfChain<C extends Chain> =
+  (typeof chainConstants)[C]['assets'][number];
+
+export type UncheckedAssetAndChain = { asset: Asset; chain: Chain };
 export type AssetAndChain = {
-  [A in Asset]: { asset: A; chain: (typeof assetChains)[A] };
-}[Asset];
+  [C in Chain]: { chain: C; asset: AssetOfChain<C> };
+}[Chain];
 
-export type UncheckedAssetAndChain = {
-  asset: Asset;
-  chain: Chain;
+export type ChainAssetMap<T> = {
+  [C in Chain]: {
+    [A in AssetOfChain<C>]: T;
+  };
 };
 
-function assertAsset(asset: string): asserts asset is Asset {
-  if (!(asset in Assets)) throw new Error('invalid asset');
+export function isValidAssetAndChain(
+  assetAndChain: UncheckedAssetAndChain,
+): assetAndChain is AssetAndChain {
+  const { asset, chain } = assetAndChain;
+  const validAssets = chainConstants[chain].assets as string[];
+
+  return validAssets.includes(asset);
 }
 
 export function assertIsValidAssetAndChain(
   assetAndChain: UncheckedAssetAndChain,
 ): asserts assetAndChain is AssetAndChain {
-  const { asset, chain } = assetAndChain;
-  assertAsset(asset);
-  if (chain !== assetChains[asset]) {
+  if (!isValidAssetAndChain(assetAndChain)) {
     throw new Error('invalid asset and chain combination');
   }
 }
+
+export const readChainAssetMap = <T>(
+  value: ChainAssetMap<T>,
+  asset: UncheckedAssetAndChain,
+): T => {
+  assertIsValidAssetAndChain(asset);
+  const chainValues = value[asset.chain];
+  return chainValues[asset.asset as keyof typeof chainValues];
+};
+
+export function getInternalAsset(asset: UncheckedAssetAndChain) {
```

Comment:
```suggestion
export function toInternalAsset(asset: UncheckedAssetAndChain) {
```

Reply:
i think i prefer get (as it's the first keyword i think of when searching)not a big fan of using `to..` unless its a chained method

Reply:
no strong opinion about this, ill do whatever you can agree on

Reply:
I don't care too much. It's just weird to me to have it be called `get` when it's a constant operation and every path could be hard-coded. To me it's more of a conversion than a lookup.